csharp/ql/src/semmle/code/csharp/Nullability.qll

```diff
@@ -0,0 +1,144 @@
+/**
+ * Provides predicates classes for representing the nullability of various C# declarations.
+ *
+ * Unlike regular types, and to preserve backwards compatibility, the nullability information is "on the side",
+ * and needs to be accessed explicitly through a `getAnnotatedType` predicate, that returns an
+ * `AnnotatedType`.
+ *
+ * Unlike regular `Type` classes, that would consider `string` and `string?` to be identical,
+ * the `AnnotatedType` of `string` and `string?` are different.
+ */
+
+import csharp
+
+module Nullability {
+  /** Holds if the flow state of the expression has determined that `e` may be null. */
+  predicate exprMayBeNull(Expr e) { nullable_flow_state(e, 2) }
+
+  /** Holds if the flow state of the expression has determined that `e` may not be null. */
+  predicate exprIsNotNull(Expr e) { nullable_flow_state(e, 1) }
+
+  private int getTypeArgumentNullability(ConstructedGeneric generic, int argument) {
+    nullable_type_arguments(generic, argument, result)
+    or
+    // Unspecified, so default to disabled (1)
+    exists(generic.getTypeArgument(argument)) and
+    not nullable_type_arguments(generic, argument, _) and
+    result = 1
+  }
+
+  private int getElementNullability(@has_nullable_annotation element) {
+    nullable_annotation(element, result)
+    or
+    // Unspecified, so default to disabled (1)
+    not nullable_annotation(element, _) and
+    result = 1
+  }
+
+  private predicate nullableElementType(@has_nullable_annotation element, Type type, int nullability) {
+    nullability = getElementNullability(element) and
+    (
+      type = element.(Assignable).getType()
+      or
+      type = element.(Callable).getReturnType()
+      or
+      type = element.(Expr).getType()
+      or
+      type = element.(ArrayType).getElementType()
+    )
+  }
+
+  private newtype TAnnotatedType =
+    TAnnotatedTypeNullability(Type type, int nullability) {
```

Comment:
Generalize this, to `string`, `string? ref`, `string ref`, `string?`
Don't just talk about Nullability.

Reply:
Represent this as a set of integer, represented as a bit field.